go/tools/gazelle/packages/package.go

```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
+		}
+		t = &p.Test
+	} else if info.isCgo || cgo && (info.category == cExt || info.category == hExt || info.category == csExt) {
+		t = &p.CgoLibrary
+	} else if info.category == goExt || info.category == sExt || !cgo && info.category == hExt {
+		t = &p.Library
+	}
+	if t != nil {
+		t.addFile(info, buildTags, platforms)
+	}
+	return nil
+}
+
+func (t *Target) addFile(info fileInfo, buildTags map[string]bool, platforms PlatformConstraints) {
+	if !info.hasConstraints() || info.checkConstraints(buildTags) {
+		t.Sources.addGenericStrings(info.name)
+		t.Imports.addGenericStrings(info.imports...)
+		t.COpts.addGenericOpts(platforms, info.copts)
+		t.CLinkOpts.addGenericOpts(platforms, info.clinkopts)
+		return
+	}
+
+	for name, tags := range platforms {
+		if info.checkConstraints(tags) {
+			t.Sources.addPlatformStrings(name, info.name)
+			t.Imports.addPlatformStrings(name, info.imports...)
+			t.COpts.addTaggedOpts(name, info.copts, tags)
+			t.CLinkOpts.addTaggedOpts(name, info.clinkopts, tags)
+		}
+	}
+}
+
+func (ps *PlatformStrings) addGenericStrings(ss ...string) {
+	ps.Generic = append(ps.Generic, ss...)
+}
+
+func (ps *PlatformStrings) addGenericOpts(platforms PlatformConstraints, opts []taggedOpts) {
+	for _, t := range opts {
+		if t.tags == "" {
+			ps.Generic = append(ps.Generic, t.opts...)
+			continue
+		}
+
+		for name, tags := range platforms {
+			if checkTags(t.tags, tags) {
+				ps.Platform[name] = append(ps.Platform[name], t.opts...)
+			}
+		}
+	}
+}
+
+func (ps *PlatformStrings) addPlatformStrings(name string, ss ...string) {
+	if ps.Platform == nil {
+		ps.Platform = make(map[string][]string)
+	}
+	ps.Platform[name] = append(ps.Platform[name], ss...)
+}
+
+func (ps *PlatformStrings) addTaggedOpts(name string, opts []taggedOpts, tags map[string]bool) {
+	if ps.Platform == nil {
+		ps.Platform = make(map[string][]string)
+	}
+	for _, t := range opts {
+		if t.tags == "" || checkTags(t.tags, tags) {
+			ps.Platform[name] = append(ps.Platform[name], t.opts...)
+		}
+	}
+}
+
+// Clean sorts and de-duplicates PlatformStrings. It also removes any
+// strings from platform-specific lists that also appear in the generic list.
+// This is useful for imports.
+func (ps *PlatformStrings) Clean() {
+	sort.Strings(ps.Generic)
+	ps.Generic = uniq(ps.Generic)
+
+	genSet := make(map[string]bool)
+	for _, s := range ps.Generic {
+		genSet[s] = true
+	}
+
+	if ps.Platform != nil {
+		for n, ss := range ps.Platform {
+			ss = remove(ss, genSet)
+			if len(ss) == 0 {
+				delete(ps.Platform, n)
+				continue
+			}
+			sort.Strings(ss)
+			ps.Platform[n] = uniq(ss)
+		}
+		if len(ps.Platform) == 0 {
+			ps.Platform = nil
+		}
+	}
+}
+
+func remove(ss []string, remove map[string]bool) []string {
+	var r, w int
+	for r, w = 0, 0; r < len(ss); r++ {
+		if !remove[ss[r]] {
+			ss[w] = ss[r]
+			w++
+		}
+	}
+	return ss[:w]
+}
+
+func uniq(ss []string) []string {
+	if len(ss) <= 1 {
+		return ss
+	}
+
+	var r, w int
```

Comment:
only w is used outside the loop, and r is the true loop control variable, so you might as well declare and init (with :=) w outside and r inside
also you can compare ss[r] to ss[w] becuase r-1 will have been copied down there anyway
also, r is used as a straight up iterator...
also, you might as well just cheat and use append rather than count and reslice, it's not going to be much worse in performance and much easier to understand
```
func uniq(ss []string) []string {
	result := ss[:0]
	previous := ""
	for _, s := range ss {
		if s != previous {
			result = append(result, s)
			previous = s
		}
	}
	return result
}
```